### DISCUSSION TITLE
Intのサブクラスを作りたい

### DISCUSSION THREAD

**Discussion by AtsushiHosaka** [lit-iphone-course/lit-iphone-course#103] [Question-original] [answered]
# やりたいこと
たとえば常に3で割ったあまりにしかならないような（2+1=0となるような）Intのサブクラスを作りたいです。しかし、Intのサブクラスがそもそも作れなさそうで、どうやったらいいのか困っています。

# やってみたこと
まずはIntを継承せずに作ってみました。
```
struct Coset {
    
    var value: Int
    let supremum: Int
    
    init(_ supremum: Int, firstValue: Int = 0) {
        
        self.value = firstValue
        self.supremum = supremum
    }
    
    static func +=(left: inout Coset, right: Int) {
        
        left.value = (left.value + right) % left.supremum
        
        if left.value < 0 {
            
            left.value += left.supremum
        }
    }
}
```

ここで、`struct Coset: Int {}`とすると、`Inheritance from non-protocol type 'Int'`と表示されました。
そこで`protocol Int {}`とプロトコルを追加してみたところ一応このエラーは消えたのですが、`let a: Int = 0`を常に`let a: Int = 0 as! Int`と書かなければならなくなりました。（`Default argument value of type 'Swift.Int' does not conform to '(プロジェクト名).Int'`とエラーが出るようになるため）

# 質問
- そもそもIntを継承するのはしないほうがいいんでしょうか？（structは基本継承できないのを調べて知りました）
- Intを継承してCosetを定義した場合、スーパークラスであるIntを呼び出すにはどうしたらいいのでしょうか？あるいは呼び出すことはできないんでしょうか？

**Comment by txxxxc** [comment, marked as answer]
深夜の返信ごめんなさい><

>そもそもIntを継承するのはしないほうがいいんでしょうか？（structは基本継承できないのを調べて知りました）

結論から言うと、継承しないほうが良いです！(できないです！)
えーえすが言ってくれている通り、Int型はstructでありstructは継承できないからです！

>Intを継承してCosetを定義した場合、スーパークラスであるIntを呼び出すにはどうしたらいいのでしょうか？あるいは呼び出すことはできないんでしょうか？

上にも書いたとおり、Intは継承できないのでスーパークラスにできないです！したがって呼び出すことはできないと考えてもらって大丈夫です！

---

>たとえば常に3で割ったあまりにしかならないような（2+1=0となるような）Intのサブクラスを作りたいです。

これを踏まえてエーエスのやりたいことを実現するには、カスタム演算子を定義するのが個人的に良いのかなと思います！
例としてはこんな感じです！↓

```
import CoreGraphics

func +(left: CGPoint, right: CGPoint) -> CGPoint {
    return CGPoint(x: left.x + right.x, y: left.y + right.y)
}

print(CGPoint(x: 1, y: 2) + CGPoint(x: 3, y: 4)) // (4.0, 6.0)
```
引用：[Swift: 演算子の定義について完全に理解した。](https://zenn.dev/kyome/articles/48f551222b339c#%E5%AE%9A%E7%BE%A9%E3%81%AE%E6%89%8B%E9%A0%86%E6%A6%82%E8%A6%81)

↑参考になると思うので良かったら使ってみてください！！
